src/main/scala/org/apache/spark/sql/kinesis/ShardSyncer.scala

```diff
@@ -198,10 +199,10 @@ private[kinesis] object ShardSyncer extends Logging {
           logDebug("Info for shardId " + shardId + " already exists")
         }
         else {
-          AddShardInfoForAncestors(shardId,
-            latestShards, initialPosition, prevShardsList, newShardsInfoMap, memoizationContext)
-          newShardsInfoMap.put(shardId,
-            new ShardInfo(shardId, initialPosition))
+          if (batchId == 0) {
```

Comment:
Why do we need this check? Why shouldnt we add shardInfo when batchId is not 0

Reply:
When the shards are scaled up or down, the shard Info of the parent shards - whose records were already consumed in the previous batches - is being added to the present batch which is resulting into the re-processing of those already processed records from TRIM_HORIZON

With this check, this ancestor info won't be added again to the current batches.

Reply:
`AddShardInfoForAncestors` is supposed to handle this. There might be some other issue. Let me look into that.

I am not sure we should add a check on batchId. Scaling shards up and down can happen at any time. We would like to include the new shards.

Reply:
New shards are added even with this check. This is just restricting the addition of ancestor info again to the batches when the shards are scaled up or down.

Reply:
@PranshuAdhikari - can you please help with sample code that can reproduce the problem.  Also can you add some unit tests for this fix.

Reply:
@itsvikramagr - 

Please find the sample code below for pushing dummy data into Kinesis stream and spark streaming code to consume data. we will face the issue when shards are scaled up or  scaled down, it reprocess the data already consumed.

sample streaming code ::

`
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.streaming.Trigger

object CheckpointingSpark {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession
      .builder
      .master("local[*]")
      .appName("TestName").getOrCreate()

    val kinesis = spark
      .readStream
      .format("kinesis")
      .option("streamName", "stream-poc")
      .option("endpointUrl", "https://kinesis.ap-south-1.amazonaws.com")
      .option("awsAccessKeyId", "#####")
      .option("awsSecretKey", "#####")
      .option("startingposition", "TRIM_HORIZON")
      .load

    kinesis.printSchema

    import spark.sqlContext.implicits._

    kinesis
      .selectExpr("CAST(data AS STRING)", "CAST(streamName AS STRING)", "CAST(partitionKey AS STRING)", "CAST(sequenceNumber AS STRING)", "CAST(approximateArrivalTimestamp AS STRING)")
      .as[(String, String, String, String, String)]
      .writeStream
      .format("console")
      .trigger(Trigger.ProcessingTime(5))
      .outputMode("append")
      .option("numRows",10000)
      .option("truncate", "false")
      .start()
      .awaitTermination()
  }
}
`
sample kinesis code to push dummy data ::

`import boto3
import json
from datetime import datetime
import time

my_stream_name = 'stream-poc'

kinesis_client = boto3.client('kinesis', region_name='ap-south-1')


def put_to_stream(id, value, timestamp):
    payload = {
                'value': value,
                'timestamp': str(timestamp),
                'id': id
              }

    print(payload)

    put_response = kinesis_client.put_record(
                        StreamName=my_stream_name,
                        Data=json.dumps(payload),
                        PartitionKey=id)
    print(put_response)

value_start_number = 0

while True:
    value = value_start_number
    timestamp = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
    print(value, timestamp, value)
    put_to_stream(str(value), value, timestamp)
    value_start_number += 1
    time.sleep(1)`

Reply:
@PranshuAdhikari  - While the issue seems to be critical, I am not sure about the solution based on the batchId check.

I have started a #62 to resolve incorrect batch-id problems along with solving serialization/deserialization. Can you try this trunk - https://github.com/itsvikramagr/kinesis-sql/tree/batch-id-issue and see if you can reproduce your problem.

Reply:
@itsvikramagr - The issue still persist with #62 .

Reply:
@PranshuAdhikari - can you email me at vagrawal at qubole.com with some more info on how you are reproducing the issue. I would also need to look at metadata saved in the checkpoint location.